lib/mix/lib/mix/tasks/app.start.ex

```diff
@@ -69,23 +69,12 @@ defmodule Mix.Tasks.App.Start do
       end
     end
 
-    # Stop Logger when starting the application as it is
-    # up to the application to decide if it should be restarted
-    # or not.
-    #
-    # Mix should not depend directly on Logger so check that it's loaded.
-    logger = Process.whereis(Logger)
-
-    if logger do
-      Logger.App.stop()
-    end
+    unless "--no-start" in args do
+      # Stop Logger if the application does not depend on it.
+      #
+      # Mix should not depend directly on Logger, that's why we first check if it's loaded.
+      if not logger_configured?() && Process.whereis(Logger), do: Logger.App.stop()
```

Comment:
Instead of relying on the configured check, we likely want to stop it whenever it isn't a child... I guess we would need to check if logger is started while we are traversing start_apps.

Reply:
>  I guess we would need to check if logger is started while we are traversing start_apps.

@josevalim if you are referring to `ensure_all_started` it relies on `Application.start/2` telling which dependent apps are not started:

https://github.com/elixir-lang/elixir/blob/d894408c91b7962a58b423b0f2c7084d1ab0f027/lib/mix/lib/mix/tasks/app.start.ex#L121-L131

Since `:logger` is already started the check cannot be done there. To know if `:logger` is not a child, we would need to stop it first.

What's that what you had in mind? 🙂

Reply:
We will need to bring the implementation of `start` over then so we can check if `logger` iss a dependency as we traverse.

Reply:
The implementation of `start` delegates to Erlang's `application` which involves `application_controller` and what not 😅 [The "started" check depends on the internal state of the `application_controller`](https://github.com/erlang/otp/blob/master/lib/kernel/src/application_controller.erl#L647-L649). It seems like a lot of work to detect if `:logger` is a dependency.

I think we could traverse the applications using `Application.spec` and check if `:logger` is listed on those. What do you think?

Reply:
Calling `spec` sounds good to me!